anx/provider/loadbalancer/loadbalancer.go

```diff
@@ -364,29 +373,58 @@ func (m mgr) checkIPCollision(ctx context.Context, ip net.IP, svc *v1.Service) e
 	return nil
 }
 
-func lbStatusFromReconcileStatus(reconStatus map[string][]uint16) *v1.LoadBalancerStatus {
-	ret := v1.LoadBalancerStatus{
-		Ingress: make([]v1.LoadBalancerIngress, 0, len(reconStatus)),
+// lbStatusFromReconcileStatus crafts a [v1.LoadBalancerStatus] out of the given service and status mapping.
+// [ipPortMap] is a map that maps from each IP address to the given ports.
+//
+// Right now, only TCP ports are supported and are therefore hardcoded.
+func lbStatusFromReconcileStatus(ipPortMap map[string][]uint16, service *v1.Service) *v1.LoadBalancerStatus {
+	// First, we're constructing a slice of unique portNumbers.
+	// This is intentionally a int32 slice to avoid the casting at a later point.
+	var portNumbers []int32
+	for _, ipPorts := range ipPortMap {
+		for _, p := range ipPorts {
+			portNumbers = append(portNumbers, int32(p))
+		}
 	}
 
-	for externalIP := range reconStatus {
-		portStatus := reconStatus[externalIP]
-		ports := make([]v1.PortStatus, 0, len(portStatus))
+	// After we constructed our slice of port numbers, we remove any duplicate elements from it.
+	slices.Sort(portNumbers)                  // sort the slice, so that compact finds duplicates
+	portNumbers = slices.Compact(portNumbers) // remove any duplicates
 
-		for _, port := range portStatus {
-			ports = append(ports, v1.PortStatus{
-				Port:     int32(port),
-				Protocol: v1.ProtocolTCP,
-			})
+	// To make use of the port numbers, we have to build a slice out of it
+	// that is compatible with our status.
+	var ports []v1.PortStatus
+	for _, p := range portNumbers {
+		// Since slices.Compact fills duplicates with the zero value, we skip them.
+		if p == 0 {
+			continue
 		}
 
-		ret.Ingress = append(ret.Ingress, v1.LoadBalancerIngress{
+		ports = append(ports, v1.PortStatus{
+			Port:     p,
+			Protocol: v1.ProtocolTCP,
+		})
+	}
+
+	// For the case that the annotation was set, we short-circuit and return the one ingress immediately.
+	// This is important, because otherwise we could return the same hostname multiple times for a dual-stack service.
+	if hostname := strings.ToLower(service.Annotations[AKEAnnotationHostname]); hostname != "" {
+		return &v1.LoadBalancerStatus{Ingress: []v1.LoadBalancerIngress{{
+			Hostname: hostname,
+			Ports:    ports,
+		}}}
+	}
+
+	// Otherwise, we construct a IP -> port mapping for each individiual IP.
+	var ingresses []v1.LoadBalancerIngress
+	for externalIP := range ipPortMap {
+		ingresses = append(ingresses, v1.LoadBalancerIngress{
 			IP:    externalIP,
 			Ports: ports,
 		})
 	}
```

Comment:
explicitly a question and not a "please change this":

The original `ipPortMap` contains the ports that were provisioned per IP. This new approach completely looses the information which address was provisioned with which ports - but I'm not sure if that is ..
1. relevant at all
2. can happen at all - if so, only if reconciliation failed I guess and then it would error out?
3. worth the effort right now

any comments on that?

Reply:
The `status` gets cleared on every reconciliation, therefore I haven't spend much time to investigate how it's behaving on errors.

Furthermore, I haven't found a case where it's relevant, so for now I ignored the fact that we lose some information in the transformation :speak_no_evil:

Reply:
gonna be good enough :tm: ^^